compilation error C2131: expression did not evaluate to a constant

Hi, I'm trying to install and use mmdetection on my Windows 10 machine (I know, that's my first problem...) and I'm at literally the last step of the process. Everything appears to have installed and compiled correctly up to this point, but when running the "pip install -v -e ." inside the mmdetection3d directory I receive the following errors for iou3d:

```
D:\Code\3DDetection\mmdetection3d\mmdet3d\ops\iou3d\src\iou3d.cpp(108): error C2131: expression did not evaluate to a constant
D:\Code\3DDetection\mmdetection3d\mmdet3d\ops\iou3d\src\iou3d.cpp(108): note: failure was caused by a read of a variable outside its lifetime
D:\Code\3DDetection\mmdetection3d\mmdet3d\ops\iou3d\src\iou3d.cpp(108): note: see usage of 'boxes_num'
D:\Code\3DDetection\mmdetection3d\mmdet3d\ops\iou3d\src\iou3d.cpp(127): error C2131: expression did not evaluate to a constant
D:\Code\3DDetection\mmdetection3d\mmdet3d\ops\iou3d\src\iou3d.cpp(108): note: failure was caused by a read of a variable outside its lifetime
D:\Code\3DDetection\mmdetection3d\mmdet3d\ops\iou3d\src\iou3d.cpp(108): note: see usage of 'boxes_num'
D:\Code\3DDetection\mmdetection3d\mmdet3d\ops\iou3d\src\iou3d.cpp(181): error C2131: expression did not evaluate to a constant
D:\Code\3DDetection\mmdetection3d\mmdet3d\ops\iou3d\src\iou3d.cpp(162): note: failure was caused by a read of a variable outside its lifetime
D:\Code\3DDetection\mmdetection3d\mmdet3d\ops\iou3d\src\iou3d.cpp(162): note: see usage of 'boxes_num'

```
In researching this particular error, it appears to be a problem with the compiler in Windows. The problem is line 108 mentioned in the above errors:

`const int col_blocks = DIVUP(boxes_num, THREADS_PER_BLOCK_NMS);`

and then its use on line 127 to initialize an array:

`unsigned long long remv_cpu[col_blocks];`

Apparently some compilers are okay with this and some, specifically in Windows apparently, are not. I do not know how to work around this problem. I have tried various things mentioned in places like StackOverflow, like setting col_blocks to be static, but the same error appears. I have also checked the issue queue for OpenPCDet, where the original code for iou3d came from, but nobody seems to have this problem. Does anyone know of a way around this?

Hi, the line which was causing error it got resolved by changing the the erroring line to dynamic array allocation at two places
`unsigned long long *remv_cpu = new unsigned long long [col_blocks];`
Now, the original error is gone, but I have a couple of errors.
Initially it says this.
```
    writing mmdet3d.egg-info\PKG-INFO
    writing dependency_links to mmdet3d.egg-info\dependency_links.txt
    writing requirements to mmdet3d.egg-info\requires.txt
    writing top-level names to mmdet3d.egg-info\top_level.txt
    reading manifest file 'mmdet3d.egg-info\SOURCES.txt'
    reading manifest template 'MANIFEST.in'
    warning: no files found matching 'mmdet3d\requirements\*.txt'
    warning: no files found matching 'mmdet3d\ops\**\*.cpp'
    warning: no files found matching 'mmdet3d\ops\**\*.cu'
    warning: no files found matching 'mmdet3d\ops\**\*.h'
    warning: no files found matching 'mmdet3d\ops\**\*.cc'
    warning: no files found matching 'mmdet3d\VERSION'
    warning: no files found matching 'mmdet3d\model_zoo.yml'
    warning: no files found matching 'mmdet3d\configs\*\*.py'
    warning: no files found matching 'mmdet3d\configs\*\*.yml'
    warning: no files found matching 'mmdet3d\tools\*.py'
    warning: no files found matching 'mmdet3d\tools\*.sh'
    warning: no files found matching 'mmdet3d\tools\*\*.py'
    warning: no files found matching 'mmdet3d\demo\*\*'
```
The `mmdet3d.ops.spconv.sparse_conv_ext` did not throw any error. But

while compiling `mmdet3d.ops.iou3d.iou3d_cuda`
```
    building 'mmdet3d.ops.iou3d.iou3d_cuda' extension
    Emitting ninja build file C:\Users\div10433\OneDrive - Esri\office-work\058-mmdet-3d\mmdetection3d\build\temp.win-amd64-3.7\Release\build.ninja...
    Compiling objects...
    Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)
    ninja: no work to do.
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.28.29333\bin\HostX86\x64\link.exe /nologo /INCREMENTAL:NO /LTCG /DLL /MANIFEST:EMBED,ID=2 /MANIFESTUAC:NO /LIBPATH:C:\Users\div10433\.conda\envs\mmcv\lib\site-packages\torch\lib "/LIBPATH:C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.0\lib/x64" /LIBPATH:C:\Users\div10433\.conda\envs\mmcv\libs /LIBPATH:C:\Users\div10433\.conda\envs\mmcv\PCbuild\amd64 "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.28.29333\ATLMFC\lib\x64" "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.28.29333\lib\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.18362.0\ucrt\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.18362.0\um\x64" c10.lib torch.lib torch_cpu.lib torch_python.lib cudart.lib c10_cuda.lib torch_cuda.lib /EXPORT:PyInit_iou3d_cuda "C:\Users\div10433\OneDrive - Esri\office-work\058-mmdet-3d\mmdetection3d\build\temp.win-amd64-3.7\Release\mmdet3d\ops\iou3d\src/iou3d.obj" "C:\Users\div10433\OneDrive - Esri\office-work\058-mmdet-3d\mmdetection3d\build\temp.win-amd64-3.7\Release\mmdet3d\ops\iou3d\src/iou3d_kernel.obj" /OUT:build\lib.win-amd64-3.7\mmdet3d\ops\iou3d\iou3d_cuda.cp37-win_amd64.pyd "/IMPLIB:C:\Users\div10433\OneDrive - Esri\office-work\058-mmdet-3d\mmdetection3d\build\temp.win-amd64-3.7\Release\mmdet3d\ops\iou3d\src\iou3d_cuda.cp37-win_amd64.lib"
       Creating library C:\Users\div10433\OneDrive - Esri\office-work\058-mmdet-3d\mmdetection3d\build\temp.win-amd64-3.7\Release\mmdet3d\ops\iou3d\src\iou3d_cuda.cp37-win_amd64.lib and object C:\Users\div10433\OneDrive - Esri\office-work\058-mmdet-3d\mmdetection3d\build\temp.win-amd64-3.7\Release\mmdet3d\ops\iou3d\src\iou3d_cuda.cp37-win_amd64.exp
    iou3d.obj : error LNK2001: unresolved external symbol "public: long * __cdecl at::Tensor::data_ptr<long>(void)const " (??$data_ptr@J@Tensor@at@@QEBAPEAJXZ)
    build\lib.win-amd64-3.7\mmdet3d\ops\iou3d\iou3d_cuda.cp37-win_amd64.pyd : fatal error LNK1120: 1 unresolved externals
    error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Professional\\VC\\Tools\\MSVC\\14.28.29333\\bin\\HostX86\\x64\\link.exe' failed with exit status 1120
    Error in atexit._run_exitfuncs:
    Traceback (most recent call last):
      File "C:\Users\div10433\.conda\envs\mmcv\lib\site-packages\colorama\ansitowin32.py", line 59, in closed
        return stream.closed
    ValueError: underlying buffer has been detached
ERROR: Command errored out with exit status 1: 'C:\Users\div10433\.conda\envs\mmcv\python.exe' -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\div10433\\OneDrive - Esri\\office-work\\058-mmdet-3d\\mmdetection3d\\setup.py'"'"'; __file__='"'"'C:\\Users\\div10433\\OneDrive - Esri\\office-work\\058-mmdet-3d\\mmdetection3d\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps Check the logs for full command output.
```

--- a/mmdet3d/ops/iou3d/src/iou3d.cpp
+++ b/mmdet3d/ops/iou3d/src/iou3d.cpp
@@ -103,7 +103,7 @@ int nms_gpu(at::Tensor boxes, at::Tensor keep,
 
   int boxes_num = boxes.size(0);
   const float *boxes_data = boxes.data_ptr<float>();
-  long *keep_data = keep.data_ptr<long>();
+  long long *keep_data = keep.data_ptr<long long>();
 
   const int col_blocks = DIVUP(boxes_num, THREADS_PER_BLOCK_NMS);
 
@@ -124,8 +124,7 @@ int nms_gpu(at::Tensor boxes, at::Tensor keep,
 
   cudaFree(mask_data);
 
-  unsigned long long remv_cpu[col_blocks];
-  memset(remv_cpu, 0, col_blocks * sizeof(unsigned long long));
+  unsigned long long* remv_cpu = new unsigned long long[col_blocks]();
 
   int num_to_keep = 0;
 
@@ -141,6 +140,7 @@ int nms_gpu(at::Tensor boxes, at::Tensor keep,
       }
     }
   }
+  delete[] remv_cpu;
   if (cudaSuccess != cudaGetLastError()) printf("Error!\n");
 
   return num_to_keep;
@@ -157,7 +157,7 @@ int nms_normal_gpu(at::Tensor boxes, at::Tensor keep,
 
   int boxes_num = boxes.size(0);
   const float *boxes_data = boxes.data_ptr<float>();
-  long *keep_data = keep.data_ptr<long>();
+  long long *keep_data = keep.data_ptr<long long>();
 
   const int col_blocks = DIVUP(boxes_num, THREADS_PER_BLOCK_NMS);
 
@@ -178,8 +178,7 @@ int nms_normal_gpu(at::Tensor boxes, at::Tensor keep,
 
   cudaFree(mask_data);
 
-  unsigned long long remv_cpu[col_blocks];
-  memset(remv_cpu, 0, col_blocks * sizeof(unsigned long long));
+  unsigned long long* remv_cpu = new unsigned long long[col_blocks]();
 
   int num_to_keep = 0;
 
@@ -195,6 +194,7 @@ int nms_normal_gpu(at::Tensor boxes, at::Tensor keep,
       }
     }
   }
+  delete[] remv_cpu;
   if (cudaSuccess != cudaGetLastError()) printf("Error!\n");
 
   return num_to_keep;

--- a/mmdet3d/ops/iou3d/src/iou3d.cpp
+++ b/mmdet3d/ops/iou3d/src/iou3d.cpp
@@ -12,6 +12,7 @@ All Rights Reserved 2019-2020.
 #include <torch/extension.h>
 #include <torch/serialize/tensor.h>
 
+#include <cstdint>
 #include <vector>
 
 #define CHECK_CUDA(x) \
@@ -103,7 +104,7 @@ int nms_gpu(at::Tensor boxes, at::Tensor keep,
 
   int boxes_num = boxes.size(0);
   const float *boxes_data = boxes.data_ptr<float>();
-  long long *keep_data = keep.data_ptr<long long>();
+  int64_t *keep_data = keep.data_ptr<int64_t>();
 
   const int col_blocks = DIVUP(boxes_num, THREADS_PER_BLOCK_NMS);
 
@@ -124,7 +125,7 @@ int nms_gpu(at::Tensor boxes, at::Tensor keep,
 
   cudaFree(mask_data);
 
-  unsigned long long* remv_cpu = new unsigned long long[col_blocks]();
+  unsigned long long *remv_cpu = new unsigned long long[col_blocks]();
 
   int num_to_keep = 0;
 
@@ -157,7 +158,7 @@ int nms_normal_gpu(at::Tensor boxes, at::Tensor keep,
 
   int boxes_num = boxes.size(0);
   const float *boxes_data = boxes.data_ptr<float>();
-  long long *keep_data = keep.data_ptr<long long>();
+  int64_t *keep_data = keep.data_ptr<int64_t>();
 
   const int col_blocks = DIVUP(boxes_num, THREADS_PER_BLOCK_NMS);
 
@@ -178,7 +179,7 @@ int nms_normal_gpu(at::Tensor boxes, at::Tensor keep,
 
   cudaFree(mask_data);
 
-  unsigned long long* remv_cpu = new unsigned long long[col_blocks]();
+  unsigned long long *remv_cpu = new unsigned long long[col_blocks]();
 
   int num_to_keep = 0;
 

--- a/mmdet3d/ops/iou3d/src/iou3d_kernel.cu
+++ b/mmdet3d/ops/iou3d/src/iou3d_kernel.cu
@@ -13,7 +13,7 @@ All Rights Reserved 2019-2020.
 
 //#define DEBUG
 const int THREADS_PER_BLOCK_NMS = sizeof(unsigned long long) * 8;
-const float EPS = 1e-8;
+__device__ const float EPS = 1e-8;
 struct Point {
   float x, y;
   __device__ Point() {}

--- a/mmdet3d/ops/paconv/src/assign_score_withk_cuda.cu
+++ b/mmdet3d/ops/paconv/src/assign_score_withk_cuda.cu
@@ -5,6 +5,7 @@
 #include <stdlib.h>
 #include <assert.h>
 #include <cmath>
+#include <cstdint>
 #include <vector>
 #include <cuda.h>
 #include <cuda_runtime.h>
@@ -49,7 +50,7 @@ __global__ void assign_score_withk_forward_kernel(const int B, const int N0, con
                                                   const float* points,
                                                   const float* centers,
                                                   const float* scores,
-                                                  const long long* knn_idx,
+                                                  const int64_t* knn_idx,
                                                   float* output) {
 
     // ----- parallel loop for B, N1, K and O ---------
@@ -82,7 +83,7 @@ __global__ void assign_score_withk_backward_points_kernel(const int B, const int
                                                           const int K, const int O, const int aggregate,
                                                           const float* grad_out,
                                                           const float* scores,
-                                                          const long long* knn_idx,
+                                                          const int64_t* knn_idx,
                                                           float* grad_points,
                                                           float* grad_centers) {
 
@@ -116,7 +117,7 @@ __global__ void assign_score_withk_backward_scores_kernel(const int B, const int
                                                           const float* grad_out,
                                                           const float* points,
                                                           const float* centers,
-                                                          const long long* knn_idx,
+                                                          const int64_t* knn_idx,
                                                           float* grad_scores) {
 
     // ----- parallel loop for B, N, K, M ---------
@@ -156,7 +157,7 @@ void assign_score_withk_forward_wrapper(int B, int N0, int N1, int M, int K, int
     const float* points_data = points.data_ptr<float>();
     const float* centers_data = centers.data_ptr<float>();
     const float* scores_data = scores.data_ptr<float>();
-    const long long* knn_idx_data = knn_idx.data_ptr<long long>();
+    const int64_t* knn_idx_data = knn_idx.data_ptr<int64_t>();
     float* output_data = output.data_ptr<float>();
 
     dim3 blocks(DIVUP(B*O*N1*K, THREADS_PER_BLOCK));
@@ -191,7 +192,7 @@ void assign_score_withk_backward_wrapper(int B, int N0, int N1, int M, int K, in
     const float* points_data = points.data_ptr<float>();
     const float* centers_data = centers.data_ptr<float>();
     const float* scores_data = scores.data_ptr<float>();
-    const long long* knn_idx_data = knn_idx.data_ptr<long long>();
+    const int64_t* knn_idx_data = knn_idx.data_ptr<int64_t>();
     float* grad_points_data = grad_points.data_ptr<float>();
     float* grad_centers_data = grad_centers.data_ptr<float>();
     float* grad_scores_data = grad_scores.data_ptr<float>();

--- a/mmdet3d/ops/paconv/src/assign_score_withk_cuda.cu
+++ b/mmdet3d/ops/paconv/src/assign_score_withk_cuda.cu
@@ -49,7 +49,7 @@ __global__ void assign_score_withk_forward_kernel(const int B, const int N0, con
                                                   const float* points,
                                                   const float* centers,
                                                   const float* scores,
-                                                  const long* knn_idx,
+                                                  const long long* knn_idx,
                                                   float* output) {
 
     // ----- parallel loop for B, N1, K and O ---------
@@ -82,7 +82,7 @@ __global__ void assign_score_withk_backward_points_kernel(const int B, const int
                                                           const int K, const int O, const int aggregate,
                                                           const float* grad_out,
                                                           const float* scores,
-                                                          const long* knn_idx,
+                                                          const long long* knn_idx,
                                                           float* grad_points,
                                                           float* grad_centers) {
 
@@ -116,7 +116,7 @@ __global__ void assign_score_withk_backward_scores_kernel(const int B, const int
                                                           const float* grad_out,
                                                           const float* points,
                                                           const float* centers,
-                                                          const long* knn_idx,
+                                                          const long long* knn_idx,
                                                           float* grad_scores) {
 
     // ----- parallel loop for B, N, K, M ---------
@@ -156,7 +156,7 @@ void assign_score_withk_forward_wrapper(int B, int N0, int N1, int M, int K, int
     const float* points_data = points.data_ptr<float>();
     const float* centers_data = centers.data_ptr<float>();
     const float* scores_data = scores.data_ptr<float>();
-    const long* knn_idx_data = knn_idx.data_ptr<long>();
+    const long long* knn_idx_data = knn_idx.data_ptr<long long>();
     float* output_data = output.data_ptr<float>();
 
     dim3 blocks(DIVUP(B*O*N1*K, THREADS_PER_BLOCK));
@@ -191,7 +191,7 @@ void assign_score_withk_backward_wrapper(int B, int N0, int N1, int M, int K, in
     const float* points_data = points.data_ptr<float>();
     const float* centers_data = centers.data_ptr<float>();
     const float* scores_data = scores.data_ptr<float>();
-    const long* knn_idx_data = knn_idx.data_ptr<long>();
+    const long long* knn_idx_data = knn_idx.data_ptr<long long>();
     float* grad_points_data = grad_points.data_ptr<float>();
     float* grad_centers_data = grad_centers.data_ptr<float>();
     float* grad_scores_data = grad_scores.data_ptr<float>();

--- a/mmdet3d/ops/voxel/src/voxelization_cpu.cpp
+++ b/mmdet3d/ops/voxel/src/voxelization_cpu.cpp
@@ -14,7 +14,8 @@ void dynamic_voxelize_kernel(const torch::TensorAccessor<T, 2> points,
                              const int NDim) {
   const int ndim_minus_1 = NDim - 1;
   bool failed = false;
-  int coor[NDim];
+  // int coor[NDim];
+  int* coor = new int[NDim]();
   int c;
 
   for (int i = 0; i < num_points; ++i) {
@@ -37,6 +38,7 @@ void dynamic_voxelize_kernel(const torch::TensorAccessor<T, 2> points,
     }
   }
 
+  delete[] coor;
   return;
 }
 

